src/org/traccar/protocol/EelinkProtocolDecoder.java

```diff
@@ -209,41 +325,72 @@ protected Object decode(
 
         ChannelBuffer buf = (ChannelBuffer) msg;
 
-        buf.skipBytes(2); // header
-        int type = buf.readUnsignedByte();
-        buf.readShort(); // length
-        int index = buf.readUnsignedShort();
+        // check that incoming protocall has correct eelink prefix
+        if (buf.readShort() != HEADER_KEY) {
+            Log.warning(new IllegalArgumentException("Invalid Header"));
+            return null;
+        }
+
+        int type = buf.readUnsignedByte(); // protocol number
 
-        if (type != MSG_GPS && type != MSG_DATA) {
-            sendResponse(channel, type, index);
+        short length = buf.readShort(); // length
+        if (buf.readableBytes() != length) { // check remaining length
+            Log.warning(new IllegalArgumentException("Packet length error"));
+            return null;
         }
 
-        if (type == MSG_LOGIN) {
+        index = buf.readUnsignedShort();
+
+        // check if we have received an old packet
+        /*int newIndex = buf.readUnsignedShort(); // current message index
```

Comment:
This part just needs a little more testing. This was why I asked a question about if the decoder was inited once for the connection or for each packet. If per connection this check would work just fine except I need to handle when it rolls over back to 0 again.